### PR TITLE
Olympus KP3R Holder

### PR DESCRIPTION
**Pull request by JExtor**

### CODE REVIEW

File: contracts/allocators/OlympusKP3RHolder.sol [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    /// @notice  Burn rKP3R and mint oKP3R
+    function oKP3Rclaim() external onlyGuardian {
+        rKP3R.claim();
+    }
+
+    /// @notice         Exercise oKP3R, pay in USDC
+    /// @param _id      ID of option to exercise / oKP3R to burn
+    /// @param _manage  Bool if managing USDC from treasury
```
Comment (ind-igo) [inline contracts/allocators/OlympusKP3RHolder.sol:157]: Comment is wrong, should be KP3R i think

**Review by ind-igo — CHANGES_REQUESTED** [review]

File: contracts/allocators/OlympusKP3RHolder.sol [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        // approve and deposit into curve
+        KP3R.approve(address(KP3RVault), _amount); 
+
+        KP3RVault.increase_amount(_amount);
+    }
+
+    /// @notice             Increases unlock time of existing lock
+    /// @param _unlockTime  Updated Timestamp at which lock will be over
+    function increaseLockTime(uint256 _unlockTime) external onlyGuardian {
+        KP3RVault.increase_unlock_time(_unlockTime);
```
Comment (lolchocotaco) [inline contracts/allocators/OlympusKP3RHolder.sol:119]: Do we need to validate the time is increasing or not in the past?

File: contracts/allocators/OlympusKP3RHolder.sol [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    /// @notice         Exercise oKP3R, pay in USDC
+    /// @param _id      ID of option to exercise / oKP3R to burn
+    /// @param _manage  Bool if managing USDC from treasury
+    function oKP3RRedeem(uint _id, bool _manage) external onlyGuardian {
+        IClaim.option memory option = rKP3R.options(_id);
+        uint _amount = option.strike;
+
+        if(_manage) { treasury.manage(address(KP3R), _amount); }
+
+        USDC.approve(address(rKP3R), _amount);
```
Comment (lolchocotaco) [inline contracts/allocators/OlympusKP3RHolder.sol:164]: Where is USDC coming into this contract?

**Review by lolchocotaco — COMMENTED** [review]
Seems reasonable! 

Can you link/write what the general flow for k3pr looks like? seems like theres a handful of different methods/usecases here that would be good to document.

File: contracts/allocators/OlympusKP3RHolder.sol (hunk 1 above)
Comment (JExtor) [reply contracts/allocators/OlympusKP3RHolder.sol:157]: Is right I believe. https://etherscan.io/address/0xEdB67Ee1B171c4eC66E6c10EC43EDBbA20FaE8e9#code

L1007

File: contracts/allocators/OlympusKP3RHolder.sol (hunk 2 above)
Comment (JExtor) [reply contracts/allocators/OlympusKP3RHolder.sol:119]: It does this in the vKP3R contract. https://etherscan.io/address/0x2FC52C61fB0C03489649311989CE2689D93dC1a2#code

L492

**Comment by JExtor** [comment]
@lolchocotaco Honestly there didn't seem to be much documentation. Here are the relevant addresses:

KP3R: 0x1ceb5cb57c4d4e2b2433641b95dd330a33185a44
vKP3R: 0x2FC52C61fB0C03489649311989CE2689D93dC1a2
rKP3R: 0xEdB67Ee1B171c4eC66E6c10EC43EDBbA20FaE8e9
oKP3R: 0xba15ab4c5b27917e19f772c9718095e248feff35
rKP3R Distributor: 0xd4260B2781e2460f49dB746112BB592ba3fb6382

**Comment by lolchocotaco** [comment]
Hey, closed the branch this PR was based off of, can you make it of of Main?

**Comment by lolchocotaco** [comment]
Hey, closed the branch this PR was based off of, can you make it of of Main?

**Comment by JExtor** [comment]
@lolchocotaco Off of main now

**Pull request closed without merging**